### PR TITLE
discovery: aws/ec2 unit tests

### PR DESCRIPTION
**Pull request by akunszt**
In it's current form this is a very minimalistic implementation of the unit test "framework" for the aws discovery. It unit tests a very short function only which makes only one call to the AWS API. So it is a good demo and people can check the logic and the implementation and possibly start a discussion.

If it passes the scrutiny then I - or any other volunteer - can implement the other, more useful tests in this or in a different PR.

Please note that my Golang knowledge is very-very basic so there could be basic issues in the code itself. I am happy to learn and improve my skills on this domain.

Related to #14310.

### CODE REVIEW

File: discovery/aws/ec2.go
Comment (rapphil) [inline discovery/aws/ec2.go:186]: I think you can still return the concrete type here. Client code can decide to use a interface or the concrete type.

Comment (akunszt) [reply discovery/aws/ec2.go:186]: I receive this when I change that back.
```
~/prometheus/discovery/aws $ go test -v
# github.com/prometheus/prometheus/discovery/aws [github.com/prometheus/prometheus/discovery/aws.test]
./ec2.go:188:10: cannot use d.ec2 (variable of type ec2iface.EC2API) as *ec2.EC2 value in return statement: need type assertion
./ec2.go:221:9: cannot use d.ec2 (variable of type ec2iface.EC2API) as *ec2.EC2 value in return statement: need type assertion
FAIL    github.com/prometheus/prometheus/discovery/aws [build failed]
```
I tried using type assertion and now it complains about functions are not implemented. Which is true as the ec2 has zillions of functions and I mocked/implemented only the ones which are called in the test.
```
ec2.EC2 does not implement ec2iface.EC2API (method AcceptAddressTransfer has pointer receiver)
```
Or maybe I misunderstood the error message. What was in your mind exactly? How to change the client code?

File: discovery/aws/ec2_test.go
Comment (rapphil) [inline discovery/aws/ec2_test.go:41]: all these utility  test functions (`RandIpv4`, `RandIpv6`....)  don't need to be public, and hence they don't need to start with a capital letter. Neither does the `EC2Data` type.

File: discovery/aws/ec2_test.go
Comment (rapphil) [inline discovery/aws/ec2_test.go:85]: instead of using this global variable, I wonder if you could create this on every test and them pass as parameter to the mock client.

e.g.:
```
type mockEC2Client struct {
	ec2iface.EC2API
        ec2Data ec2Data
}
```

File: discovery/aws/ec2_test.go
Comment (rapphil) [inline discovery/aws/ec2_test.go:237]: nit: please notice that all your tests end up asserting these same conditions.  

I think you can refactor the tests for the `refresh` function specifically to be table driven. https://go.dev/wiki/TableDrivenTests

The idea is that you build a table where each row contains a test name, the input data for the tests and optionally some data used in assertions. This way it is easier to add new test cases.

**Comment by rapphil** [comment]
Thank you for this PR. This is great! 

I placed some comments. The comment about table driven development is a nice to have.

File: discovery/aws/ec2.go
Comment (akunszt) [reply discovery/aws/ec2.go:186]: Thank you. I marking this thread as resolved to keep the PR easy to read.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go]: (Hopefully) fixed in https://github.com/prometheus/prometheus/pull/14364/commits/b72467f305c9286142d099c37eef4214b3b41ee8.

What about the `GenerateTest*` functions. Should I make those private as well?

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go]: I hope https://github.com/prometheus/prometheus/pull/14364/commits/1774d8354499c017a8e5772a5183a6fe6390af2d is what you meant.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go]: I did something similar in https://github.com/prometheus/prometheus/pull/14364/commits/5b66987fb14c0a20507fb03a69af36ae79ec5e39 but I still had to use generator functions to create the input and the expected output. But the list of tests is more manageable this way. I think.

**Review by simonpasquier — COMMENTED** [review]
Thanks! My personal opinion is that randomized inputs make the tests harder to read/follow but I can't say that it's a hard blocker for me. Let see what other maintainers have to say :)

**Comment by akunszt** [comment]
@simonpasquier I applied the changes you requested.

I usually prefer to use semi-random test data to avoid any accidental assumptions from the human side. The semi-random means that I always pay attention to make the generated data distinguishable from each other (wherever is possible). During a failed test people can see what "fields" are failed, not just two random strings. In this case I added a separate prefix for every different kind of AWS resource ID.

Of course if the consensus is to use static data then I will rewrite according to that. Earlier I planned to migrate ec2.go to the AWS SDK v2 (it uses the obsolete v1 right now) but strictly after there are tests to verify the results.

File: discovery/aws/ec2_test.go
Comment (bboreham) [inline discovery/aws/ec2_test.go:1]: 2024?

File: discovery/aws/ec2_test.go
Comment (bboreham) [inline discovery/aws/ec2_test.go:113]: Who will fix it?  What happens if we don't?

File: discovery/aws/ec2_test.go
Comment (bboreham) [inline discovery/aws/ec2_test.go:382]: Suggest `TestEC2DiscoveryRefresh`, as a more helpful name when something random fails.

**Review by bboreham — COMMENTED** [review]
Generally fine thank you; couple more nits from me.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go:1]: 2024? Already? Did I miss 9 years??!! Did anything interesting happened?

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go:382]: Fixed the name.

Are the other names are fine? Don't you see the test file path or something else unique in the output of the Github action?

Comment (bboreham) [reply discovery/aws/ec2_test.go]: You do, but have to hunt around.  Also if I want to file a bug about this test I would have to copy-paste parts of the path to form a unique name.

In this example, it's somewhat visible in the 3rd line, and also in the 33rd line:
```
--- FAIL: TestStreamReadEndpoint (0.10s)
    read_handler_test.go:296: 
        	Error Trace:	/__w/prometheus/prometheus/storage/remote/read_handler_test.go:296
        	Error:      	Not equal: 
        	            	expected: []*prompb.ChunkedReadResponse{(*prompb.ChunkedReadResponse)(0xc00369fd40), (*prompb.ChunkedReadResponse)(0xc00369fd80), (*prompb.ChunkedReadResponse)(0xc00369fdc0), (*prompb.ChunkedReadResponse)(0xc00369fe00), (*prompb.ChunkedReadResponse)(0xc00369fe40), (*prompb.ChunkedReadResponse)(0xc00369fe80)}
        	            	actual  : []*prompb.ChunkedReadResponse{(*prompb.ChunkedReadResponse)(0xc00369f9c0), (*prompb.ChunkedReadResponse)(0xc00369fa40), (*prompb.ChunkedReadResponse)(0xc00369fac0), (*prompb.ChunkedReadResponse)(0xc00369fb40), (*prompb.ChunkedReadResponse)(0xc00369fbc0), (*prompb.ChunkedReadResponse)(0xc00369fc80)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -162,4 +162,4 @@
        	            	       Type: (prompb.Chunk_Encoding) 1,
        	            	-      Data: ([]uint8) (len=15) {
        	            	-       00000000  00 01 80 f4 ee 06 40 c7  70 00 00 00 00 00 00     |......@.p......|
        	            	+      Data: ([]uint8) (len=14) {
        	            	+       00000000  00 01 80 f4 ee 06 40 c7  70 00 00 00 00 00        |......@.p.....|
        	            	       },
        	            	@@ -344,4 +344,4 @@
        	            	       Type: (prompb.Chunk_Encoding) 1,
        	            	-      Data: ([]uint8) (len=15) {
        	            	-       00000000  00 01 80 e8 dd 0d 40 d7  70 00 00 00 00 00 00     |......@.p......|
        	            	+      Data: ([]uint8) (len=14) {
        	            	+       00000000  00 01 80 e8 dd 0d 40 d7  70 00 00 00 00 00        |......@.p.....|
        	            	       },
        	Test:       	TestStreamReadEndpoint
FAIL
FAIL	github.com/prometheus/prometheus/storage/remote	110.963s
```

File: discovery/aws/ec2_test.go
Comment (machine424) [inline discovery/aws/ec2_test.go:319]: I think that instead of using setters/field assignments, composite literal initialization would be more readable, WDYT?

**Review by machine424 — COMMENTED** [review]
Thanks for this.

Although I agree with you on the advantages of monkey testing, I agree with Simon and think that readability and reproducibility are two things we should not give up.

Being able to see the final expected meta labels makes it easier to reason about the test and gives you confidence in it; Having that built and injected under the hood may make you question whether that logic needs to be tested as well, whether it covers a case you want to add, etc.

Also, even if randomized inputs can help discover some edge cases, we’ll still need to accompany the fix with static cases to avoid regression.

As most of the test inputs for other SD are hardcoded, I think it’s good to keep consistency.

I think, as a first step, let’s go with static input, then we can introduce some randomization later to help us discover edge cases if needed.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go]: AWS SDK expects pointers, even on boolean or int64 values. IDK why.
As you requested I modified the tests to use only static data but I had to add ugly addrX functions to get pointers (that's how Go calls them) on literals.
So this code is gone by now.

**Review by machine424 — COMMENTED** [review]
way clearer (for me at least), thanks.
We're inches away from merging this, just some propositions.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go]: They were used in the mock function but I managed to rewrite that and use only `azToAZID`.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go]: I did as you requested. In the future we can put all those `xxxptr` functions into a common package. If it worth the effort, of course.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go:94]: Done.

File: discovery/aws/ec2_test.go
Comment (akunszt) [reply discovery/aws/ec2_test.go]: Done.

**Review by machine424 — APPROVED** [review]
Thanks again for taking care of this.
This will be really helpful.
lgtm, let's see if the others have something to add.

**Comment by machine424** [comment]
Let's give this a try.

**Pull request merged**
